crates/red_knot/src/lint.rs

```diff
@@ -57,6 +62,108 @@ pub(crate) fn lint_lines(source: &str, diagnostics: &mut Vec<String>) {
     }
 }
 
+#[tracing::instrument(level = "debug", skip(db))]
+pub(crate) fn lint_semantic<Db>(db: &Db, file_id: FileId) -> Diagnostics
+where
+    Db: SemanticDb + HasJar<SemanticJar>,
+{
+    let storage = &db.jar().lint_semantic;
+
+    storage.get(&file_id, |file_id| {
+        let source = db.source(*file_id);
+        let parsed = db.parse(*file_id);
+        let symbols = db.symbol_table(*file_id);
+
+        let context = SemanticLintContext {
+            file_id: *file_id,
+            source,
+            parsed,
+            symbols,
+            db,
+            diagnostics: RefCell::new(Vec::new()),
+        };
+
+        lint_unresolved_imports(&context);
+
+        Diagnostics::from(context.diagnostics.take())
+    })
+}
+
+fn lint_unresolved_imports(context: &SemanticLintContext) {
+    for (symbol, definition) in context.symbols().all_definitions() {
```

Comment:
On a typical large real-world module, we might have literally thousands of definitions (every assignment statement), out of which tens are imports. Maybe we should index the import definitions separately? That could also unify `definitions` and `dependencies`.

Reply:
That's a good point. Let's solve this when we implement the real rule ;) I thought I was already pretty clever by using definitions instead of iterating over the entire AST again